checker/jdk/nullness/src/java/util/EnumMap.java

```diff
@@ -0,0 +1,839 @@
+             // to note on toArray in Collection.java
+         })
+        public <T> @Nullable T[] toArray(@Nullable T[] a) {
+            int size = size();
+            if (a.length < size)
+                a = (T[])java.lang.reflect.Array
+                    .newInstance(a.getClass().getComponentType(), size);
+            if (a.length > size)
+                a[size] = null;
+            return (T[]) fillEntryArray(a);
+        }
+        @SuppressWarnings("nullness:argument.type.incompatible") // Value returned by unmaskNull will be
+        // of type V for mapped value
+        private Object[] fillEntryArray(Object[] a) {
+            int j = 0;
+            for (int i = 0; i < vals.length; i++)
+                if (vals[i] != null)
+                    a[j++] = new AbstractMap.SimpleEntry<>(
+                        keyUniverse[i], unmaskNull(vals[i]));
+            return a;
+        }
+    }
+
+    private abstract class EnumMapIterator<T> implements Iterator<T> {
+        // Lower bound on index of next element to return
+        int index = 0;
+
+        // Index of last returned element, or -1 if none
+        int lastReturnedIndex = -1;
+
+        public boolean hasNext() {
+            while (index < vals.length && vals[index] == null)
+                index++;
+            return index != vals.length;
+        }
+
+        public void remove() {
+            checkLastReturnedIndex();
+
+            if (vals[lastReturnedIndex] != null) {
+                vals[lastReturnedIndex] = null;
+                size--;
+            }
+            lastReturnedIndex = -1;
+        }
+
+        private void checkLastReturnedIndex() {
+            if (lastReturnedIndex < 0)
+                throw new IllegalStateException();
+        }
+    }
+
+    private class KeyIterator extends EnumMapIterator<K> {
+        public K next() {
+            if (!hasNext())
+                throw new NoSuchElementException();
+            lastReturnedIndex = index++;
+            return keyUniverse[lastReturnedIndex];
+        }
+    }
+
+    private class ValueIterator extends EnumMapIterator<V> {
+        @SuppressWarnings("nullness:return.type.incompatible") // Value returned by unmaskNull will be
+        // of type V for mapped value
+        public V next() {
+            if (!hasNext())
+                throw new NoSuchElementException();
+            lastReturnedIndex = index++;
+            return unmaskNull(vals[lastReturnedIndex]);
+        }
+    }
+
+    private class EntryIterator extends EnumMapIterator<Map.Entry<K,V>> {
+        private @Nullable Entry lastReturnedEntry = null;
+
+        public Map.Entry<K,V> next() {
+            if (!hasNext())
+                throw new NoSuchElementException();
+            lastReturnedEntry = new Entry(index++);
+            return lastReturnedEntry;
+        }
+
+        @SuppressWarnings("nullness:dereference.of.nullable") // super.remove throws a NullPointerException
```

Comment:
I would use `@AssumeAssertion` to put this comment on the line after the `super.remove` call.  I think it will be easier to read and understand there.